extensions/in-memory/negotiation-store-memory/src/main/java/org/eclipse/dataspaceconnector/negotiation/store/memory/InMemoryContractNegotiationStore.java

```diff
@@ -110,9 +110,20 @@ public void delete(String processId) {
     public @NotNull List<ContractNegotiation> nextForState(int state, int max) {
         return lockManager.readLock(() -> {
             var set = stateCache.get(state);
-            return set == null ? Collections.emptyList() : set.stream()
+
+            List<ContractNegotiation> toBeLeased = set == null ? Collections.emptyList() : set.stream()
                     .sorted(Comparator.comparingLong(ContractNegotiation::getStateTimestamp)) //order by state timestamp, oldest first
                     .limit(max)
+                    .collect(toList());
+
+            stateCache.compute(state, (key, value) -> {
+                if (value != null) {
+                    value.removeAll(toBeLeased);
+                }
+                return value;
+            });
+
+            return toBeLeased.stream()
```

Comment:
Wouldn't a similiar type of lease also be needed for the `CosmosContractNegotiationStore`?

Reply:
Already done in #724